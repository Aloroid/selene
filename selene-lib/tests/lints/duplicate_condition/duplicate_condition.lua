print(x) -- just x, doesn't simplify
print(x and x) -- simplifies to x
print(x and x or y) -- simplifies to x or y
print(x or x and y) -- simplifies to x
print(x or y or x) -- simplifies to x or y
print(not x and not x) -- simplifies to not x
print(x and y or z) -- doesn't simplify

print(3 and 2) -- simplifies to 3, but lint doesn't support that right now

-- The QMC implementation only allows 32 conditions.
-- This is so unlikely that I don't really care what output it gives,
-- just that it does not panic (and doesn't lie).
print(
    x1
    and x2
    and x3
    and x4
    and x5
    and x6
    and x7
    and x8
    and x9
    and x10
    and x11
    and x12
    and x13
    and x14
    and x15
    and x16
    and x17
    and x18
    and x19
    and x20
    and x21
    and x22
    and x23
    and x24
    and x25
    and x26
    and x27
    and x28
    and x29
    and x30
    and x31
    and x32
    and x33
    and x34
    and x35
    and x36
)

-- none of these are solvable, as call() can have side effects
print(call() and call())
print(call() and call() or y)
print(call() or call() and y)
print(call() or y or call())
print(not call() and not call())

while foo and foo or bar do end
while foo or foo and bar do end
repeat until not foo and not foo
repeat until x.y or x.y
