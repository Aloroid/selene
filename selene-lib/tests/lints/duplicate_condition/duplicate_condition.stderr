error[duplicate_condition]: condition can be simplified
  ┌─ duplicate_condition.lua:2:7
  │
2 │ print(x and x) -- simplifies to x
  │       ^^^^^^^ can be simplified to `x`

error[duplicate_condition]: condition can be simplified
  ┌─ duplicate_condition.lua:3:7
  │
3 │ print(x and x or y) -- simplifies to x or y
  │       ^^^^^^^^^^^^ can be simplified to `x or y`

error[duplicate_condition]: condition can be simplified
  ┌─ duplicate_condition.lua:4:7
  │
4 │ print(x or x and y) -- simplifies to x
  │       ^^^^^^^^^^^^ can be simplified to `x`

error[duplicate_condition]: condition can be simplified
  ┌─ duplicate_condition.lua:5:7
  │
5 │ print(x or y or x) -- simplifies to x or y
  │       ^^^^^^^^^^^ can be simplified to `x or y`

error[duplicate_condition]: condition can be simplified
  ┌─ duplicate_condition.lua:6:7
  │
6 │ print(not x and not x) -- simplifies to not x
  │       ^^^^^^^^^^^^^^^ can be simplified to `not x`

error[duplicate_condition]: condition can be simplified
   ┌─ duplicate_condition.lua:61:7
   │
61 │ while foo and foo or bar do end
   │       ^^^^^^^^^^^^^^^^^^ can be simplified to `foo or bar`

error[duplicate_condition]: condition can be simplified
   ┌─ duplicate_condition.lua:62:7
   │
62 │ while foo or foo and bar do end
   │       ^^^^^^^^^^^^^^^^^^ can be simplified to `foo`

error[duplicate_condition]: condition can be simplified
   ┌─ duplicate_condition.lua:63:14
   │
63 │ repeat until not foo and not foo
   │              ^^^^^^^^^^^^^^^^^^^ can be simplified to `not foo`

error[duplicate_condition]: condition can be simplified
   ┌─ duplicate_condition.lua:64:14
   │
64 │ repeat until x.y or x.y
   │              ^^^^^^^^^^ can be simplified to `x.y`

